pkg/proc/process_tree.go

```diff
@@ -0,0 +1,162 @@
+package proc
+
+import (
+	"bytes"
+	"cmp"
+	"os"
+	"path/filepath"
+	"slices"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/castai/kvisor/pkg/system"
+)
+
+type Process struct {
+	PID  PID
+	PPID PID
+	Args []string
+	// StartTime since boot start
+	StartTime time.Duration
+	FilePath  string
+}
+
+// SnapshotProcessTree records a snappshot of the current process tree in the PID namespace of the
+// given targetPID. This is done by iterating over files exposed from the `/proc` filesystem.
+func (p *Proc) SnapshotProcessTree(targetPID PID) ([]Process, error) {
+	targetPIDString := pidToString(targetPID)
+	targetPath := filepath.Join(targetPIDString, "root", "proc")
+
+	entires, err := p.procFS.ReadDir(targetPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// This will always overshoot with memory, but still better than not pre-allocating.
+	processes := make([]Process, 0, len(entires))
+
+	for _, de := range entires {
+		// We only care about processes, hence we test for numbers only folders.
+		if !de.IsDir() || !numOnlyName(de.Name()) {
+			continue
+		}
+
+		pid, err := parsePID(de.Name())
+		if err != nil {
+			return nil, err
+		}
+
+		if pid == 2 {
```

Comment:
what is the significance of pid == 2?

Reply:
I saw the comment below
```
			// All processes under PID 2 (kthreadd) are kernel threads we do not want to display
```
maybe duplicate it here?

Reply:
Another option would be to extract this PID into a well named constant, for example, `kthreadPid` and update the comment that we're not interested in kthreads here.